core/store/store.go

```diff
@@ -34,3 +34,28 @@ type KVStore interface {
 
 // Iterator is an alias db's Iterator for convenience.
 type Iterator = dbm.Iterator
+
+type CoreIterator interface {
+	// Domain returns the start (inclusive) and end (exclusive) limits of the iterator.
+	// CONTRACT: start, end readonly []byte
+	Domain() ([]byte, []byte, error)
+
+	// Valid returns whether the current iterator is valid. Once invalid, the Iterator remains
+	// invalid forever.
+	Valid() (bool, error)
+
+	// Next moves the iterator to the next key in the database, as defined by order of iteration.
+	// If Valid returns false, this method will error.
+	Next() error
```

Comment:
I'm wondering how scanning of an iterator can be done without being too cumbersome API wise,  but overall looks good to me

Reply:
It would be good to see some usage examples

Reply:
I guess sth like this?

```golang
for it.Valid() {
  key := it.Key()

  if err := it.Next(); err != nil {
    return err
  }
}
```

Reply:
previously there's also some cases that the iterator already has an error after construction, that case seems not covered by current interface.

Reply:
the idea here was to make the original method return an error because people could forget to check the error. If others don't think this is necessary we can close it and leave the interface as is. I don't have a strong presence